### ISSUE TITLE
Compatibility Report - tch-rs 0.14.0 with PyTorch 2.8.0 and CUDA 12.8

### ISSUE THREAD

**Issue by rbardyla** [LaurentMazare/tch-rs#978]
[TEST_REPORT.md](https://github.com/user-attachments/files/22127376/TEST_REPORT.md)

## Summary

Successfully tested tch-rs 0.14.0 with PyTorch 2.8.0 and CUDA 12.8. The library works well but requires `LIBTORCH_BYPASS_VERSION_CHECK=1` and shows some deprecation warnings.

## Environment

- **tch-rs Version**: 0.14.0
- **PyTorch Version**: 2.8.0+cu128
- **CUDA Version**: 12.8
- **GPUs Tested**: 
  - NVIDIA GeForce RTX 5080 (16GB) - Primary testing
  - NVIDIA GeForce RTX 2070 SUPER (8GB) - Backward compatibility
- **OS**: Linux
- **Rust Version**: 1.83.0

## What Works ✅

1. **Basic tensor operations** - All examples run successfully
2. **CUDA support** - GPU operations work correctly
3. **Neural networks** - Training and inference on GPU works
4. **CPU-GPU transfers** - Zero data loss in transfers
5. **Matrix operations** - 1000x1000 matmul in ~32ms on RTX 2070S

### 1. Version Check
tch-rs expects PyTorch 2.7.1 but I have 2.8.0. Works with:
```bash
export LIBTORCH_BYPASS_VERSION_CHECK=1
```

### 2. Deprecation Warnings
```
libtch/torch_api.cpp:193: warning: 'at::autocast::is_enabled()' is deprecated
libtch/torch_api.cpp:201: warning: 'at::autocast::set_enabled(bool)' is deprecated
```

The deprecated functions should be replaced with:
- `at::autocast::is_autocast_enabled(at::kCUDA)`
- `at::autocast::set_autocast_enabled(at::kCUDA, enabled)`

## Test Code

Created comprehensive GPU test (see attached `test_gpu.rs`):
```rust
use tch::{nn, nn::Module, Device, Tensor};

fn main() {
    // Test CUDA availability
    let cuda_available = tch::Cuda::is_available();
    println!("CUDA available: {}", cuda_available);
    
    // Test GPU operations
    let device = Device::Cuda(0);
    let a = Tensor::randn(&[1000, 1000], (tch::Kind::Float, device));
    let b = Tensor::randn(&[1000, 1000], (tch::Kind::Float, device));
    let c = a.matmul(&b);
    
    // Test neural network
    let vs = nn::VarStore::new(device);
    let net = nn::seq()
        .add(nn::linear(&vs.root() / "layer1", 784, 256, Default::default()))
        .add_fn(|xs| xs.relu())
        .add(nn::linear(&vs.root() / "layer2", 256, 10, Default::default()));
    
    let input = Tensor::randn(&[32, 784], (tch::Kind::Float, device));
    let output = net.forward(&input);
    println!("Output shape: {:?}", output.size());
}
```

## Suggestions

1. **Update PyTorch version support** to 2.8.0
2. **Fix deprecated autocast API** calls in `torch_api.cpp`
3. **Document the `LIBTORCH_BYPASS_VERSION_CHECK`** environment variable in README

## Questions

1. Are there plans to officially support PyTorch 2.8.0?
2. Would you accept a PR to fix the deprecated API warnings?
3. Is there a recommended way to handle version mismatches for users who need newer PyTorch?

## Full Test Report

See attached `TEST_REPORT.md` for detailed test results and performance metrics.

---

Thank you for maintaining this excellent library! Despite the minor compatibility issues, tch-rs works great with the latest PyTorch and CUDA versions.

use tch::{nn, nn::Module, Device, Tensor};

fn main() {
    println!("Testing tch-rs GPU functionality");
    println!("==================================");
    
    // Check CUDA availability
    let cuda_available = tch::Cuda::is_available();
    let cudnn_available = tch::Cuda::cudnn_is_available();
    let device_count = tch::Cuda::device_count();
    
    println!("CUDA available: {}", cuda_available);
    println!("CUDNN available: {}", cudnn_available);
    println!("Number of CUDA devices: {}", device_count);
    
    if !cuda_available {
        println!("CUDA is not available, skipping GPU tests");
        return;
    }
    
    // Select GPU device
    let device = Device::Cuda(0);
    println!("\nUsing device: {:?}", device);
    
    // Create tensors on GPU
    println!("\n1. Creating tensors on GPU...");
    let a = Tensor::randn(&[1000, 1000], (tch::Kind::Float, device));
    let b = Tensor::randn(&[1000, 1000], (tch::Kind::Float, device));
    println!("   Created two 1000x1000 tensors on GPU");
    
    // Perform matrix multiplication on GPU
    println!("\n2. Performing matrix multiplication on GPU...");
    let start = std::time::Instant::now();
    let c = a.matmul(&b);
    let elapsed = start.elapsed();
    println!("   Matrix multiplication completed in {:?}", elapsed);
    
    // Test moving tensors between CPU and GPU
    println!("\n3. Testing tensor movement between CPU and GPU...");
    let cpu_tensor = Tensor::randn(&[100, 100], (tch::Kind::Float, Device::Cpu));
    println!("   Created tensor on CPU");
    
    let gpu_tensor = cpu_tensor.to(device);
    println!("   Moved tensor to GPU");
    
    let back_to_cpu = gpu_tensor.to(Device::Cpu);
    println!("   Moved tensor back to CPU");
    
    // Verify the values are preserved
    let diff = (&cpu_tensor - &back_to_cpu).abs().max().double_value(&[]);
    println!("   Max difference after round trip: {:.10}", diff);
    
    // Test neural network on GPU
    println!("\n4. Testing neural network on GPU...");
    let vs = nn::VarStore::new(device);
    let net = nn::seq()
        .add(nn::linear(&vs.root() / "layer1", 784, 256, Default::default()))
        .add_fn(|xs| xs.relu())
        .add(nn::linear(&vs.root() / "layer2", 256, 128, Default::default()))
        .add_fn(|xs| xs.relu())
        .add(nn::linear(&vs.root() / "layer3", 128, 10, Default::default()));
    
    let input = Tensor::randn(&[32, 784], (tch::Kind::Float, device));
    let output = net.forward(&input);
    println!("   Neural network forward pass completed");
    println!("   Output shape: {:?}", output.size());
    
    // Note: Memory stats not available in this version
    println!("\n5. GPU Memory Usage:");
    println!("   Memory stats API not available in tch 0.14.0");
    println!("   (Use nvidia-smi for memory monitoring)");
    
    println!("\n✅ All GPU tests passed successfully!");
    println!("==================================");
}

5080 test 
use tch::{Device, Tensor, nn, nn::OptimizerConfig, Kind};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🎮 RTX 5080 Blackwell Test");
    println!("=========================");
    
    // Check if CUDA is available
    println!("CUDA available: {}", tch::Cuda::is_available());
    println!("CUDA device count: {}", tch::Cuda::device_count());
    
    if !tch::Cuda::is_available() {
        println!("❌ CUDA not available - falling back to CPU");
        return Ok(());
    }
    
    // Test RTX 5080 specifically
    let device = Device::Cuda(0);
    println!("🚀 Using device: {:?}", device);
    
    // Test basic tensor operations
    println!("\n📊 Testing tensor operations on RTX 5080...");
    let x = Tensor::randn(&[1024, 1024], (Kind::Float, device));
    let y = Tensor::randn(&[1024, 1024], (Kind::Float, device));
    
    let start = std::time::Instant::now();
    let z = x.matmul(&y);
    let duration = start.elapsed();
    
    println!("✅ Matrix multiplication (1024x1024): {:?}", duration);
    println!("   Result shape: {:?}", z.size());
    println!("   Memory usage: {:.1} MB", (z.numel() * 4) as f32 / 1024.0 / 1024.0);
    
    // Test neural network operations
    println!("\n🧠 Testing neural network on RTX 5080...");
    let vs = nn::VarStore::new(device);
    let root = vs.root();
    
    let linear1 = nn::linear(&root / "linear1", 1024, 512, Default::default());
    let linear2 = nn::linear(&root / "linear2", 512, 256, Default::default());
    let linear3 = nn::linear(&root / "linear3", 256, 10, Default::default());
    
    let input = Tensor::randn(&[32, 1024], (Kind::Float, device)); // Batch of 32
    
    let start = std::time::Instant::now();
    let out = input
        .apply(&linear1)
        .relu()
        .apply(&linear2)
        .relu()
        .apply(&linear3);
    let duration = start.elapsed();
    
    println!("✅ Forward pass (3-layer MLP): {:?}", duration);
    println!("   Input shape: {:?}", input.size());
    println!("   Output shape: {:?}", out.size());
    
    // Test memory allocation
    println!("\n💾 Testing memory allocation...");
    let large_tensor = Tensor::zeros(&[4096, 4096], (Kind::Float, device));
    println!("✅ Allocated 64MB tensor successfully");
    println!("   Tensor shape: {:?}", large_tensor.size());
    
    // Test compute capability
    println!("\n🔧 RTX 5080 Blackwell Features:");
    println!("   Compute Capability: 12.0 (sm_120)");
    println!("   CUDA Version: 12.8+");
    println!("   Memory: 16GB GDDR7");
    
    println!("\n🎉 RTX 5080 test completed successfully!");
    Ok(())
}

**Comment by Ynkcc** [comment]
I encountered the same problem.
